x/perp/v2/keeper/dnr_test.go

```diff
@@ -97,3 +97,110 @@ func TestUserVolumes(t *testing.T) {
+	startBlockTime := time.Now()
+
+	exchangeFee := sdk.MustNewDecFromStr("0.0010")           // 0.1%, default fee taken from CreateCustomMarketAction
+	globalFeeDiscount := sdk.MustNewDecFromStr("0.0005")     // 0.05%
+	fauxGlobalFeeDiscount := sdk.MustNewDecFromStr("0.0006") // 0.06%
+	customFeeDiscount := sdk.MustNewDecFromStr("0.0002")     // 0.02%
+	fauxCustomFeeDiscount := sdk.MustNewDecFromStr("0.0003") // 0.03%
+
+	tests := TestCases{
+		TC("user does not have any past epoch volume: no discount applies").
+			Given(
+				DnREpochIs(1),
+				CreateCustomMarket(
+					pairBtcNusd,
+					WithPricePeg(sdk.OneDec()),
+					WithSqrtDepth(sdk.NewDec(100_000)),
+				),
+				SetBlockNumber(1),
+				SetBlockTime(startBlockTime),
+
+				FundAccount(alice, sdk.NewCoins(sdk.NewCoin(denoms.NUSD, positionSize.AddRaw(1000)))),
+				FundModule(types.PerpEFModuleAccount, sdk.NewCoins(sdk.NewCoin(denoms.NUSD, sdk.NewInt(100_000_000)))),
+			).
+			When(
+				DnREpochIs(1),
+			).
+			Then(
+				MarketOrderFeeIs(exchangeFee, alice, pairBtcNusd, types.Direction_LONG, sdk.NewInt(10_000), sdk.OneDec(), sdk.ZeroDec()),
+			),
+		TC("user has past epoch volume: no discount applies").
+			Given(
+				DnREpochIs(2),
+				CreateCustomMarket(
+					pairBtcNusd,
+					WithPricePeg(sdk.OneDec()),
+					WithSqrtDepth(sdk.NewDec(100_000)),
+				),
+				SetBlockNumber(1),
+				SetBlockTime(startBlockTime),
+
+				FundAccount(alice, sdk.NewCoins(sdk.NewCoin(denoms.NUSD, positionSize.AddRaw(1000)))),
+				FundModule(types.PerpEFModuleAccount, sdk.NewCoins(sdk.NewCoin(denoms.NUSD, sdk.NewInt(100_000_000)))),
+			).
+			When(
+				SetGlobalDiscount(fauxGlobalFeeDiscount, sdk.NewInt(50_000)),
+				SetGlobalDiscount(globalFeeDiscount, sdk.NewInt(100_000)),
+				SetCustomDiscount(alice, fauxCustomFeeDiscount, sdk.NewInt(50_000)),
+				SetCustomDiscount(alice, customFeeDiscount, sdk.NewInt(100_000)),
+				SetPreviousEpochUserVolume(alice, sdk.NewInt(10_000)), // lower than 50_000
+			).
+			Then(
+				MarketOrderFeeIs(exchangeFee, alice, pairBtcNusd, types.Direction_LONG, sdk.NewInt(10_000), sdk.OneDec(), sdk.ZeroDec()),
+			),
+		TC("user has past epoch volume: custom discount applies").
+			Given(
+				DnREpochIs(2),
+				CreateCustomMarket(
+					pairBtcNusd,
+					WithPricePeg(sdk.OneDec()),
+					WithSqrtDepth(sdk.NewDec(100_000)),
+				),
+				SetBlockNumber(1),
+				SetBlockTime(startBlockTime),
+
+				FundAccount(alice, sdk.NewCoins(sdk.NewCoin(denoms.NUSD, positionSize.AddRaw(1000)))),
+				FundModule(types.PerpEFModuleAccount, sdk.NewCoins(sdk.NewCoin(denoms.NUSD, sdk.NewInt(100_000_000)))),
+			).
+			When(
+				SetGlobalDiscount(sdk.MustNewDecFromStr("0.0004"), sdk.NewInt(50_000)),
+				SetGlobalDiscount(fauxGlobalFeeDiscount, sdk.NewInt(100_000)),
+				SetCustomDiscount(alice, fauxCustomFeeDiscount, sdk.NewInt(50_000)),
+				SetCustomDiscount(alice, customFeeDiscount, sdk.NewInt(100_000)),
+				SetPreviousEpochUserVolume(alice, sdk.NewInt(100_000)),
```

Comment:
Would love to see a test case where the `previousEpochUserVolume` !== the threshold value (e.g. `100_001`), to test that the descending iteration on the collection works as expected.